auto_gptq/nn_modules/qlinear/qlinear_cuda.py

```diff
@@ -267,10 +268,10 @@ def forward(self, x: torch.Tensor):
                     g_idx_i = self.g_idx[i*num_dim:(i+1)*num_dim]
                     weights.append(scale_i[g_idx_i.long()] * (weight_i - zeros_i[g_idx_i.long()]))
                 weights = torch.cat(weights,dim=1)
-            out = torch.matmul(x.to(weights.dtype), weights)
-        out = out.half().reshape(out_shape)
+            out = torch.matmul(x, weights)
+        out = out.to(dtype=weights.dtype).reshape(out_shape)
```

Comment:
This seems broken in master at the moment:

  File "/opt/miniconda3/envs/text-gen-gptq/lib/python3.10/site-packages/auto_gptq/nn_modules/qlinear/qlinear_cuda.py", line 272, in forward
    out = out.to(dtype=weights.dtype).reshape(out_shape)
UnboundLocalError: local variable 'weights' referenced before assignment

Reply:
Hi @chuyqa, thanks for pointing out this issue. I have added a fix here: https://github.com/PanQiWei/AutoGPTQ/pull/390.